runtime/service.go

```diff
@@ -236,6 +240,72 @@ func (s *service) monitorState(ctx context.Context, id, exec_id string, pid uint
 	}
 }
 
+func (s *service) proxyStdio(ctx context.Context, stdin, stdout, stderr string, CID uint32) {
+	go proxyIn(ctx, stdin, CID, internal.StdinPort)
+	go proxyOut(ctx, stdout, CID, internal.StdoutPort)
+	go proxyOut(ctx, stderr, CID, internal.StderrPort)
+}
+
+func proxyIn(ctx context.Context, path string, CID, port uint32) {
+	if path != "" {
+		reader, err := fifo.OpenFifo(ctx, path, syscall.O_RDONLY|syscall.O_NONBLOCK, 0700)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("error opening fifo")
+			return
+		}
+		conn, err := vsock.Dial(CID, port)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("unable to dial agent vsock")
+		}
+		buf := make([]byte, internal.DefaultBufferSize)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				n, err := reader.Read(buf)
+				n, werr := conn.Write(buf[:n])
```

Comment:
Why not to use `io.CopyBuffer(conn, reader, buf) / CopyN / LimitReader`  instead?
This will simplify error handling as well.